### ISSUE TITLE
feat: Automatic query invalidations

### ISSUE THREAD

**Issue by jtmarmon** [trpc/trpc#5264]
### Describe the feature you'd like to request

I would love the ability to define invalidations in my query/mutation definitions, so that queries can be automatically invalidated without having to call it manually on the client. 

### Describe the solution you'd like to see

**Example**
Given a mutation called `updateUser` and a query called `getUser`. In the client right now, everywhere you call `updateUser`, you need to remember to invalidate `getUser` to prevent staleness.

Instead, it'd be great if this could be done automatically for you if your definition looks like so:
```js
const usersRouter = router({
   getUser: procedure.input(...).query(() => ...),
   updateUser: procedure.input(...).invalidates('getUser').mutation(...)
});
```

### Describe alternate solutions

n/a

### 👨‍👧‍👦 Contributing

- [ ] 🙋‍♂️ Yes, I'd be down to file a PR implementing this feature!

**Comment by Glinkis** [comment]
This seems like it wouldn't be possible without breaking the server/client code separation, since the router is defined on the server, and **only** the ttpes are available to the client.

**Comment by KATT** [comment]
I don't think this would require a breaking considering our [RPC-spec](https://trpc.io/docs/rpc#successful-response) has the ability to return more data

```js
{
  result: {
    data: TOutput; // output from procedure
  }
}
```

A mutation could return e.g.

```js
{
  result: {
    data: TOutput; // output from procedure
  },
  invalidates: [
    ['post', 'list'], // invalidate post.list
    ['post', 'byId', 1], // invalidate post id 1
  ]
}
```

**Comment by KATT** [comment]
It will be tricky in the `@trpc/react-query` package though with how/when we actually could automatically trigger invalidations and how loading state would be, but not impossible.

Would a mutation stay "loading" until everything has been invalidated? Timing matters here if you do e.g. a redirect as a result of a mutation - you'd likely want the invalidation happen _after_ a route change - think in the case of deleting a post you're viewing as an example. Would likely need to be configurable.

**Comment by KATT** [comment]
My recommendation is to [Invalidate full cache on every mutation](https://trpc.io/docs/client/react/useUtils#invalidate-full-cache-on-every-mutation) until you actually have a need for anything else

**Comment by DawidWraga** [comment]
This would be a really great addition to TRPC. @KATT perhaps [zenstacks implementation](https://zenstack.dev/docs/the-complete-guide/part4/invalidation) could be a useful source of inspiration

**Comment by hunterwilhelm** [comment]
Here's an approach for type safe invalidation without circular type inference problems. *

Since invalidations are side effects, a contextual approach would look something like this.

```ts
const mockDb: Record<number, string> = { 1: "Hello World" };
const todoRouter = createTRPCRouter({
  updateTodo: procedurePublic
    .input(z.object({ id: z.number(), text: z.string() }))
    .mutation(async ({ input }) => {
      mockDb[input.id] = input.text;
      todoRouter.getTodo.invalidate({ id: input.id }); // adds information to the context to send to the client to invalidate the cache
    }),
  getTodo: procedurePublic
    .input(z.object({ id: z.number() }))
    .query(async ({ input }) => {
      return mockDb[input.id];
    }),
});
```

*: Using the types from a different route in the return statement would cause a `is referenced in its own initializer.` error.

**Comment by svaraborut** [comment]
The proposal is interesting, but imho it breaks the separation of concerns as under the proposed solution the server should take care of the cache that only exists on the fe. 

In the [docs](https://trpc.io/docs/client/react/useUtils) I see how to automatically invalidate the whole cache, but it is not clear how to **automatically** invalidate the router wrapping a mutation.

I think that allowing a mutation to automatically invalidate all the sibling queries (within the same router) will cover the majority of common use cases. 

How can this be achieved via client implementation?

**Comment by jtmarmon** [comment]
I guess you could argue it breaks separation of concerns, but BE sending caching info to the frontend is typical (e.g Etag). I think you might be able to hack this together on the client with a custom Link, though

**Comment by GreccoOliva-Franco** [comment]
Hi, can we at least expose the invalidation method directly through the `<route>.useMutation()`? This should not be a problem on Separation of Concerns and the DX is better avoiding typos

Current format:
```
const cache = api.useUtils()
const method = api.<router>.<method>.useMutation({
  onSuccess: async () => { await cache.<router>.<method>.invalidate() }
})
```

Proposal:
```
const method = api.<router>.<method>.useMutation({ ... })

// on some action
method.mutate({
  ...,
  onSuccess: async () => { await method.invalidate() }
})
```

Again, this way you avoid typos, gain readability in you code and Separation of Concerns is untouched
